src/core/scope.ml

```diff
@@ -156,279 +151,312 @@ let scope : mode -> sig_state -> env -> p_term -> tbox = fun md ss env t ->
+  | (P_Prod(_,_), M_Patt) ->
+      fatal t.pos "Dependent products are not allowed in a pattern."
+  | (P_Prod(xs,b), _) -> scope_binder md ss _Prod env xs (Some(b))
+  | (P_LLet(x,xs,a,t,u), M_Term(_))
+  | (P_LLet(x,xs,a,t,u), M_URHS(_))
+  | (P_LLet(x,xs,a,t,u), M_RHS(_)) ->
+      let a = scope_binder md ss _Prod env xs a in
+      let t = scope_binder md ss _Abst env xs (Some(t)) in
+      let v = new_tvar x.elt in
+      let u = scope md ss (Env.add v a (Some(t)) env) u in
+      if not (Bindlib.occur v u) then
+        wrn x.pos "Useless let-binding ([%s] not bound)." x.elt;
+      _LLet a t (Bindlib.bind_var v u)
+  | (P_LLet(_), M_LHS(_)) ->
+      fatal t.pos "Let-bindings are not allowed in a LHS."
+  | (P_LLet(_), M_Patt) ->
+      fatal t.pos "Let-bindings are not allowed in a pattern."
+  | (P_NLit(n), _) ->
+      let sym_z = _Symb (Builtin.get ss t.pos "0")
+      and sym_s = _Symb (Builtin.get ss t.pos "+1") in
+      let rec unsugar_nat_lit acc n =
+        if n <= 0 then acc else unsugar_nat_lit (_Appl sym_s acc) (n-1) in
+      unsugar_nat_lit sym_z n
+  | (P_Wrap(t), _) -> scope md ss env t
+  | (P_Expl(_), _) -> fatal t.pos "Explicit argument not allowed here."
 
-(** [scope ?exp ss env t] turns a parser-level term [t] into an actual term.
+(** [scope expo ss env t] turns a parser-level term [t] into an actual term.
     The variables of the environment [env] may appear in [t]. The signature
     state [ss] is used to handle module aliasing according to [find_qid]. If
-    [?exp] is {!constructor:Public}, then the term mustn't contain any private
+    [expo] is {!constructor:Public}, then the term mustn't contain any private
     subterms. *)
 let scope_term :
      Tags.expo -> sig_state -> env -> meta IntMap.t Lazy.t -> p_term -> term =
   fun expo ss env sgm t ->
   let m = Stdlib.ref StrMap.empty in
   Bindlib.unbox (scope (M_Term(m, sgm, expo)) ss env t)
 
+(** [scope_params expo ss env sgm t] turns a parser-level desugarised
+    definition [t] into an actual term. Parameters [expo], [ss], [env] and
+    [sgm] have the same behaviour as in {!val:scope_term}. Unlike
+    {!val:scope_term}, this function does not print warnings when binders
+    are constant because parameters of definition may appear solely in the
+    type or in the definition. These checks have to be performed
+    afterwards. *)
+let scope_params :
+     Tags.expo -> sig_state -> env -> meta IntMap.t Lazy.t -> p_term -> term =
+  fun expo ss env sgm t ->
+  let m = Stdlib.ref StrMap.empty in
+  let md = M_Term(m, sgm, expo) in
+  if Timed.(!log_enabled) then log_scop "%a" Pretty.term t;
+  (* Extract the spine. *)
+  let p_head, _ = get_pratt_args ss env t in
+  let scope_b cons xs t =
+    let warn = false in
+    Bindlib.unbox (scope_binder ~warn md ss cons env xs (Some t)) in
+  match p_head.elt with
+  | P_Abst(xs,t) -> scope_b _Abst xs t
+  | P_Prod(xs,t) -> scope_b _Prod xs t
+  | _ -> scope_term expo ss env sgm t
```

Comment:
I've changed your code here: in the case of constant definitions, the assert false would trigger. Perhaps you intended to check before calling `scope_params` that the list of parameters is not empty?

Reply:
Otherwise we can just split into two functions, one that scopes types and the other that scopes terms. In fact I prefer this alternative.

Reply:
I don't know exactly why but your proposal was failing to pass all tests.